### PR TITLE
Remove tSafeWatcher and vChannel when release

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

issues: #12259

### CODE REVIEW

**Comment by binbinlv** [comment]
This is an urgent PR for critical issue, so set CI-passed first.

@bigsheeper please do not forget to add ut after this, thanks.

File: internal/querynode/task.go [hunk 1]
```diff
@@ -847,6 +892,21 @@ func (r *releasePartitionsTask) Execute(ctx context.Context) error {
...
+				// and remove the corresponding channel in collection
+				qc, err := r.node.queryService.getQueryCollection(r.req.CollectionID)
+				if err != nil {
+					return err
+				}
+				err = qc.removeTSafeWatcher(channel)
+				if err != nil {
+					return err
+				}
+				sCol.removeVChannel(channel)
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querynode/task.go]: fixed

File: internal/querynode/task.go [hunk 2]
```diff
@@ -715,6 +715,21 @@ func (r *releaseCollectionTask) releaseReplica(replica ReplicaInterface, replica
 			if err != nil {
 				log.Warn(err.Error())
 			}
+			_, err = r.node.tSafeReplica.getTSafe(channel)
```
Comment (xiaofan-luan) [inline internal/querynode/task.go:718]: why do we not move the check into removeTSafeWatcher

File: internal/querynode/query_collection.go [hunk 3]
```diff
@@ -193,15 +193,32 @@ func (q *queryCollection) addTSafeWatcher(vChannel Channel) error {
...
 	)
-	go q.startWatcher(q.tSafeWatchers[vChannel].watcherChan())
+	go q.startWatcher(q.tSafeWatchers[vChannel].watcherChan(), q.tSafeWatchers[vChannel].closeCh)
 	return nil
 }
 
-// TODO: add stopWatcher(), add close() to tSafeWatcher
-func (q *queryCollection) startWatcher(channel <-chan bool) {
+func (q *queryCollection) removeTSafeWatcher(channel Channel) error {
+	q.tSafeWatchersMu.Lock()
```
Comment (xiaofan-luan) [inline internal/querynode/query_collection.go:201]: add log here

File: internal/querynode/task.go (hunk 2 above)
Comment (bigsheeper) [reply internal/querynode/task.go]: because we should remove both tSafeWatcher and collection's channel,
how about returning `isRemoved` in removeTSafe?

**Comment by bigsheeper** [comment]
> This is an urgent PR for critical issue, so set CI-passed first.
> 
> @bigsheeper please do not forget to add ut after this, thanks.

I'v added ut, so remove `ci-passed` label

File: internal/querynode/task.go (hunk 2 above)
Comment (xiaofan-luan) [reply internal/querynode/task.go]: > 

that would make more sense to me~

File: internal/querynode/query_collection.go (hunk 3 above)
Comment (bigsheeper) [reply internal/querynode/query_collection.go:201]: added

File: internal/querynode/task.go (hunk 2 above)
Comment (bigsheeper) [reply internal/querynode/task.go]: okay

File: internal/querynode/collection.go [hunk 4]
```diff
@@ -170,15 +185,21 @@ OUTER:
 
 // getPChannels get physical channels of collection
 func (c *Collection) getPDeltaChannels() []Channel {
+	c.channelMu.RLock()
+	defer c.channelMu.RUnlock()
 	return c.pDeltaChannels
```
Comment (xiaofan-luan) [inline internal/querynode/collection.go:190]: shoud do copy before return to avoid race

File: internal/querynode/collection.go [hunk 5]
```diff
@@ -170,15 +185,21 @@ OUTER:
...
 }
 
 func (c *Collection) getVDeltaChannels() []Channel {
+	c.channelMu.RLock()
+	defer c.channelMu.RUnlock()
 	return c.vDeltaChannels
```
Comment (xiaofan-luan) [inline internal/querynode/collection.go:196]: same here, should do copy before release

File: internal/querynode/tsafe_replica.go [hunk 6]
```diff
@@ -114,15 +115,16 @@ func (t *tSafeReplica) removeTSafe(vChannel Channel) error {
 	if t.tSafes[vChannel].ref == 0 {
 		safer, err := t.getTSaferPrivate(vChannel)
```
Comment (xiaofan-luan) [inline internal/querynode/tsafe_replica.go:116]: remove the err logic for getTSaferPrivate,we don't need to return err in removeTSafe, only return true or false

File: internal/querynode/collection.go (hunk 4 above)
Comment (bigsheeper) [reply internal/querynode/collection.go]: fixed

File: internal/querynode/collection.go (hunk 5 above)
Comment (bigsheeper) [reply internal/querynode/collection.go]: fixed

File: internal/querynode/tsafe_replica.go (hunk 6 above)
Comment (bigsheeper) [reply internal/querynode/tsafe_replica.go:116]: fixed

**Pull request merged**